### DISCUSSION TITLE
How to call JetBrains action from a Vimscript function?

### DISCUSSION THREAD

**Discussion by JustBarnt** [JetBrains/ideavim#1033] [Q&A] [answered]
I'm trying to figure out how to call a JetBrains action such as `EditorChooseLookupItem` or `InsertInlineCompletionAction`. I'm sure there isn't currently a way to tell whether there is an inline suggestion active or if the completion menu is active, so I thought I would just run both and see what happens

My goal is trying to make `<C-Y>` work for inserting from the autocomplete menu or when IntelliSense suggests a line based on context. 

Currently I have based of reading the `README` where it says `:action {action_id}` would work in the Ex Commandline.
```vim
function! ChooseCompletionAction()
    :action EditorChooseLookupItem<CR>
    :action InsertInlineCompletionAction<CR>
endfunction

imap <C-y> <Esc>:call ChooseCompletionAction()<CR>a
```

**Comment by lippfi** [comment]
It's not a mapping, but a Vim script command, so you don't need the `<CR>` in `:action EditorChooseLookupItem<CR>`
You also don't need the leading `:`, but it shouldn't break your script

**Reply by JustBarnt** [reply]
I'll give this a try. Would you also happen to know why I can't seem to bind something like `EditorDown` to `<C-J>`? I have a `sethandler <C-J> a:vim` call so in all modes it lets ideavim handle the key but I still can't seem to make anything work with it.

I can easily bind `<C-K>` to something like `EditorUp` and it works just fine though. I've also double check that there are no other actions bound to `<C-J>` though Rider.

**Reply by JustBarnt** [reply]
This still doesn't seem to work so maybe this is just not something that's achievable then?
``` vim
function! ChooseCompletionAction()
    action EditorChooseLookupItem
    action InsertInlineCompletionAction
endfunction

imap <C-y> <Esc>:call ChooseCompletionAction()<CR>a
```

**Reply by OptimusCrime** [reply]
Are you sure that the actions are correct? The following function worked fine (both actions are performed):

```viml
function! Foo()
  action Annotate
  action NewScratchFile
endfunction

:map <leader>7 :call Foo()<CR>
```

**Reply by citizenmatt** [reply]
Your map is typing `<Escape>` to exit insert mode while there's a completion window open. Wouldn't that Escape hide the lookup?

I wonder if `normal <Action>(EditorChooseLookupItem)` would work?

**Reply by JustBarnt** [reply]
@OptimusCrime Yes I have set both `EditorChooseLookupItem` and `InsertInlineCompletionAction` as indivdual keymaps and they work. I am just trying to set my `ideavim` config as close to what I can with in my actually `nvim` configuration and I am used to pressing `<C-y>` to accept the completion item. This works just fine for `EditorChooseLookupItem`, but it doesn't work when Rider is trying to give me an inline completion, which I would also like to be able to press `<C-y>` to accept as well intead of having to press `<TAB>`. So I had the idea of just writing a function that would call them both in hopes it would try one then other but it seems like it just quits if it can't call the other successfully maybe?

@citizenmatt That's a good point about it hiding the completion popup, so that's probably why it doesn't work when trying to call from my function. It sounds like there is currently no way for me to achieve this since it seems like there is also no variable or editor state I can check to see if the is currently a Popup open or a inline completion action available. 

But I thank you both for the time you took to respond.

**Reply by lippfi** [reply]
What about something like:
```
map <expr> <C-y> (&ide =~? 'rider') ? "<Action>(EditorChooseLookupItem)" : "<Action>(InsertInlineCompletionAction)"
```
or more manageable way:
```
if &ide =~? 'rider'
  " rider mappings
else
  " other mappings
endif
```

It's not really a good approach to execute two different things expecting only one of them will work
But anyway something more staightforward like `map <C-y> <Action>(EditorChooseLookupItem)<Action>(InsertInlineCompletionAction)` should also work, but I'm not sure

Update:
If Rider has both of these actions and it is not a naming issue, please let me know if the last option is working. I'll try to fix it if it isn't

**Reply by JustBarnt** [reply]
@lippfi Rider does have both of those actions. I currently have them bound like this since I haven't been able to get them both to work under one keymap. 

```vim
" Currently set as two keymaps while I try to find a solution to mapping to <C-y> only.
imap <C-y> <action>(EditorChooseLookupItem) "completes popup items when the completion menu is open
imap <A-y> <action>(InsertInlineCompletionAction) "Completes the inline completions that show through virtual text
```

**Reply by citizenmatt** [reply]
Make sure you don't leave those comments at the end of the `imap` statements in your actual `ideavimrc` - they're not comments and actually part of the mapping!

**Reply by JustBarnt** [reply]
@citizenmatt I don't but thank you for mentioning that! I just wanted to put them there in the snippet for clarity! 😄

**Reply by lippfi** [reply, marked as answer]
@JustBarnt so what about the `map <C-y> <Action>(EditorChooseLookupItem)<Action>(InsertInlineCompletionAction)` option? Is it working?

**Reply by JustBarnt** [reply]
@lippfi I will try that, I have tried `imap <C-y> <Action>(EditorChooseLookupItem)<Action>(InsertInlineCompletionAction)` but not `map`, but that wouldn't make much of a difference would it? It would just map it in all modes right?

**Reply by lippfi** [reply]
@JustBarnt oh no, sorry, it should be `imap` for insert mode, you are right
Didn't `imap` work? I've tried other action sequence as a test, and both of them were executed

**Reply by JustBarnt** [reply]
@lippfi Yeah `imap` didn't work with those two actions at all. I have also gotten other sequences as well to work and both were executed. So it seems like this just isn't feasible, it almost as if when one fails to run it just exit the sequence. So I may just have to deal with my current mappings of <C-y> and <A-y> haha.

**Reply by lippfi** [reply]
I'll try to debug a have a look tomorrow

**Reply by JustBarnt** [reply]
@lippfi Oh awesome! Well thank you very much!

**Reply by JustBarnt** [reply]
@lippfi Update, you're last suggestion actually does work. I was looking at my config and I noticed that I was doing 
```vim
map <C-y> <Action>(EditorChooseLookupItem) <Action>(InsertInlineCompletionAction)
```
and not 
```vim
map <C-y> <Action>(EditorChooseLookupItem)<Action>(InsertInlineCompletionAction)
```

So I tried removing the space and it works just fine!
